machine/v1/types_nutanixprovider.go

```diff
@@ -154,6 +166,161 @@ type NutanixResourceIdentifier struct {
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'Name' ?  has(self.name) : !has(self.name)",message="name configuration is required when type is Name, and forbidden otherwise"
+// +union
+type NutanixGPU struct {
+	// type is the identifier type to use for this GPU resource.
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	Type NutanixGPUIdentifierType `json:"type"`
+
+	// deviceID is the id of the GPU entity.
+	// +optional
+	// +unionMember
+	DeviceID *int64 `json:"deviceID,omitempty"`
+
+	// name is the GPU name
+	// +optional
+	// +unionMember
+	Name *string `json:"name,omitempty"`
+}
+
+// NutanixStorageContainerReferenceType is an enumeration of different storage_container reference types.
+// +kubebuilder:validation:Enum:=UUID;Name
+type NutanixStorageContainerReferenceType string
+
+const (
+	// NutanixStorageContainerReferenceUUID is a reference type identifying the storage_container object by UUID.
+	NutanixStorageContainerReferenceUUID NutanixStorageContainerReferenceType = "UUID"
+
+	// NutanixStorageContainerReferenceName is a reference type identifying the storage_container object by Name.
+	NutanixStorageContainerReferenceName NutanixStorageContainerReferenceType = "Name"
+
+	// NutanixStorageContainerReferenceURL is a reference type identifying the storage_container object by URL.
+	NutanixStorageContainerReferenceURL NutanixStorageContainerReferenceType = "URL"
+)
+
+// NutanixStorageContainerReference references to a storage_container object.
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'UUID' ?  has(self.uuid) : !has(self.uuid)",message="uuid configuration is required when type is UUID, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'Name' ?  has(self.name) : !has(self.name)",message="name configuration is required when type is Name, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'URL' ?  has(self.url) : !has(self.url)",message="url configuration is required when type is URL, and forbidden otherwise"
+// +union
+type NutanixStorageContainerReference struct {
+	// type is the reference type for reference to the storage_container object.
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	Type NutanixStorageContainerReferenceType `json:"type"`
+
+	// uuid is the UUID of the storage_container object. It cannot be empty if the type is UUID.
+	// +optional
+	// +unionMember
+	UUID *string `json:"uuid,omitempty"`
+
+	// name is the name of the storage_container object. It cannot be empty if the type is Name.
+	// +optional
+	// +unionMember
+	Name *string `json:"name,omitempty"`
+
+	// url is the source URL of the storage_container object. It cannot be empty if the type is URL.
+	// +optional
+	// +unionMember
+	URL *string `json:"url,omitempty"`
```

Comment:
For each of these, the godoc should explain the format of the strings that go into them, max length, min length, valid characters, any pattern matching

Reply:
This code is refactored, since currently we can only support identify the storage_container object by its UUID.